### PR TITLE
[FEAT] Approximate quantile aggregation (pulled into main)

### PR DESCRIPTION
**Pull request by jaychia**
Puts the finishing touches on #2076

### CODE REVIEW

**Review by colin-ho — COMMENTED** [review]
Overall lookin really good! A few high level things:
- Add the expression to the docs
- Do we want to expose top level `approx_percentile` expressions for this? (on df and grouped df) so users can do `df.approx_percentiles("values")` or `df.group_by('group').approx_percentiles("values")`

File: daft/expressions/expressions.py [hunk 1]
```diff
@@ -434,6 +434,21 @@ def sum(self) -> Expression:
         expr = self._expr.sum()
         return Expression._from_pyexpr(expr)
 
+    def approx_percentiles(self, percentiles: builtins.float | builtins.list[builtins.float]) -> Expression:
+        """Calculates the approximate percentile(s) for a float column
```
[the remark this answers is not in the document]
Comment (jaychia) [reply daft/expressions/expressions.py:438]: We should actually expose this functionality to support all ordered columns (numeric, temporal, strings...)

Then we can leverage this for our sorts. Currently our sorts perform its own bespoke solution for sampling for boundaries before repartitioniong, but we can leverage `approx_percentiles` to generate those boundaries for us!

File: src/daft-dsl/src/expr.rs [hunk 2]
```diff
@@ -52,10 +57,30 @@ pub enum Expr {
     },
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct ApproxPercentileParams {
+    pub child: ExprRef,
+    pub percentiles: Vec<f64>,
```
[the remark this answers is not in the document]
Comment (jaychia) [reply src/daft-dsl/src/expr.rs:63]: Oh awesome! Didn't know we already had this.

File: daft/expressions/expressions.py [hunk 3]
```diff
@@ -434,6 +434,21 @@ def sum(self) -> Expression:
...
+        Args:
+            percentiles: the percentile(s) at which to find approximate values at. Can be provided as a single
+                float or a list of floats.
+
+        Returns:
+            Expression: a new expression representing the approximate percentile(s). If `percentiles` was a single float,
+                this will be a new `Float64` expression. If `percentiles` was a list of floats, this will be a new expression
+                with type: `FixedSizeList[Float64, len(percentiles)]`.
+        """
+        expr = self._expr.approx_percentiles(percentiles)
```
[the remark this answers is not in the document]
Comment (jaychia) [reply daft/expressions/expressions.py:449]: I think we can link to the paper for the implementation here

File: src/daft-dsl/src/python.rs [hunk 4]
```diff
@@ -295,6 +301,28 @@ impl PyExpr {
         Ok(self.expr.clone().sum().into())
     }
 
+    pub fn approx_percentiles(&self, percentiles: ApproxPercentileInput) -> PyResult<Self> {
```
[the remark this answers is not in the document]
Comment (jaychia) [reply src/daft-dsl/src/python.rs:304]: I was hoping to actually differentiate between if a user passes in `[0.5]` vs `0.5`. In the former case maybe we should return a size-1 fixed size list, and in the latter just return floats.

I didn't end up implementing this logic though, but this API lets us do that (we can keep track of a flag `force_fixed_size_list_output: bool` which we flip to True if the user passes in `[0.5]`)

[the remark this answers is not in the document]
Comment (jaychia) [reply src/daft-dsl/src/python.rs:304]: I pushed a new commit to make this the case, and also unit test this behavior.

**Review by colin-ho — APPROVED** [review]
LGTM!

**Comment by jaychia** [comment]
FYI @maxime-petitjean !

**Pull request merged**